### DISCUSSION TITLE
I hate CORS, please help 🙇🏻‍♂️

### DISCUSSION THREAD

**Discussion by BSBjorn** [Diizzayy/nuxt-graphql-client#260] [Q&A]
I'm sorry, but I'm really bad with CORS... Just never really understand how to work with it.

I have a local dev environment (localhost:3000) where I try to do a graphql mutation for an external API. I set some header for authentication like this:
```
runtimeConfig: {
    public: {
      "graphql-client": {
        clients: {
          default: {
            host: process.env.GQL_HOST,
          },
          search: {
            host: "https://api.crystallize.com/hageland-dev/search",
          },
          orders: {
            host: "https://api.crystallize.com/hageland-dev/orders",
            headers: {
              "X-Crystallize-Access-Token-Id":
                process.env.CRYSTALLIZE_ACCESS_TOKEN_ID,
              "X-Crystallize-Access-Token-Secret":
                process.env.CRYSTALLIZE_ACCESS_TOKEN_SECRET,
            },
          },
        },
      },
    },
  }
```
But when I try to query it, this is the respons I get... 

![Screenshot 2022-11-23 at 13 12 36](https://user-images.githubusercontent.com/19681090/203544199-7f8fd2cb-6216-4a41-abdb-9631255d5e7c.png)

I'm calling the API like this: 
```
async function submitOrder() {
  useGqlCors({ credentials: "same-origin" });
  //useGqlCors({ credentials: "Access-Control-Allow-Origin" });
  const { createOrder } = await GqlCreateOrder();
  console.log(createOrder);
}
```
I understand this is probably my own fault and lack of understanding, but I just want to learn how to work with this... 
Thanks in advance for any help! <3

**Comment by Diizzayy** [comment]
@BSBjorn Given that your api is hosted on an external domain and that it does infact allow for requests from `localhost:3000`, you have to use the `include` credentials as opposed to `same-origin`.

`useGqlCors({ credentials: 'include' })`

**Reply by BSBjorn** [reply]
Thanks for the reply! I moved to logic to a composable, but I still get the same error when I try the query? Is there something wrong with my code? 

```
export async function createCrystallizeOrder(order) {
  //const orderObject = JSON.stringify(order);
  //console.log(orderObject);
  useGqlCors({ credentials: "include" });
  const { response } = await GqlCreateOrder();
  console.log(response);
}
´´´

**Reply by Diizzayy** [reply]
@BSBjorn There's also a possibility that what you're facing here is partly due to the request in the browser being unauthorized, in which case you'd also be hit with cors errors as it would fail as well.

Have you ensured that the client side request is being sent with the proper headers for auth?

Also feel free to reach out to me privately on discord (Diizzayy#1964) where we may be able to communicate faster.

**Reply by BSBjorn** [reply]
Yeah, you where right, the problem was that the headers where not set appropriately, I added a plugin to the browser that modifies the header, and now I get to create mutations on the API. I can now see that the headers are added correctly

![Screenshot 2022-11-23 at 21 39 57](https://user-images.githubusercontent.com/19681090/203642272-1cbe4101-23f9-4543-b967-2fa45324fe19.png)


Allthough my problem is initially solved, I feel that this is some workaround that is not fit for production, and I'm not sure why the headers where not set in the first place, as they are defined in `nuxt.config.ts` ? 
```
orders: {
            host: "https://api.crystallize.com/hageland-dev/orders",
            headers: {
              "X-Crystallize-Access-Token-Id":
                process.env.CRYSTALLIZE_ACCESS_TOKEN_ID,
              "X-Crystallize-Access-Token-Secret":
                process.env.CRYSTALLIZE_ACCESS_TOKEN_SECRET,
            },
          },
```
Would love to have you help me understand this, I'll reach out on Discord tomorrow.

**Reply by michael-mb** [reply]
Same issue https://github.com/CrystallizeAPI/boilerplates/issues/228

**Reply by michael-mb** [reply]
@BSBjorn Which plugin did you use please ?

**Reply by Diizzayy** [reply]
@michael-mb Would you be willing to allow me to use your repo to further debug and resolve this issue?

**Reply by michael-mb** [reply]
Yes yes it would help me a lot
